### ISSUE TITLE
Add a spinner while loading fullsize image

### ISSUE THREAD

**Issue by jimblue** [kingdido999/zooming#75]
Hi there!

Actually when clicking on an image, `Zooming` doesn't provide any visual information to let the user know a better resolution image will show when loaded.

This can be a problem on devices with poor connection as users might close the `Zoomed` image before the full size image is displayed.

To let users know something better is coming, what do you think about adding a div with a css spinner while the full size image is loading.

Thanks for the good work on this awesome library btw 😄 !!

Cheers

**Comment by kingdido999** [comment]
Yeah this is a good idea. However, I want to be careful about making it an opt-in feature because some people might not needed it. Maybe we can introduce two new event hooks `onImageLoading` and `onImageLoaded`, then user can add/remove loading indicator with their own styling. I'll experiment with it and see how it goes :)

**Comment by jimblue** [comment]
Cool! New event hooks totally make because of your points and regarding customisation too!

**Comment by kingdido999** [comment]
The above two event hooks have been added to the latest release `2.1.0`:
- [onImageLoading](https://desmonding.me/zooming/docs/#/configuration?id=onimageloading)
- [onImageLoaded](https://desmonding.me/zooming/docs/#/configuration?id=onimageloaded)

But I haven't figured out the most appropriate way to indicate the loading status yet. One thing I'm sure about is that we need to set a certain time threshold for the indicator to only show up in slow internet connection, otherwise it blinks which is a bit annoying.

**Comment by migueldemoura** [comment]
> One thing I'm sure about is that we need to set a certain time threshold for the indicator to only show up in slow internet connection or the first fetch, otherwise it blinks which is a bit annoying.

[Turbolinks](https://github.com/turbolinks/turbolinks/blob/583df0a767c7d06fbd4a0061ae58a1f1b0ee7177/src/turbolinks/controller.coffee#L16) does this with a 500ms delay before showing the progress bar. I doubt this is based on any UX study, but I quite like the value.

**Comment by jimblue** [comment]
Hey @kingdido999,

Thanks for the new release!
I've wrote a promise function to check image load properly.
I use it to lazyload image, it's pretty solid, you can try it.

```js
/**
 * Method to fetch image from a source
 *
 * @param {URL} imgSource
 * @returns {Promise} with image Node
 */
const fetchImage = imgSource => {
  // Return a promise to fetch the image
  return new Promise((resolve, reject) => {
    // Create a new image
    const imgElement = new Image()

    // Define events callbacks
    const onLoad = () => {
      event.currentTarget.removeEventListener('load', onLoad)
      resolve(imgElement)
    }
    const onError = () => {
      event.currentTarget.removeEventListener('error', onError)
      reject(Error(`Error while loading image at ${imgSource}`))
    }

    // Listen to load and error event on image
    imgElement.addEventListener('load', onLoad)
    imgElement.addEventListener('error', onError)

    // Set image source
    imgElement.src = imgSource
  })
}
```

You can use it like so:

```js
fetchImage(imgSource).then(() => { console.log('image loaded') })
```

**Comment by jimblue** [comment]
About time threshold, as usage will probably be different for each case, I suggest you to simply add an exemple to the documentaion.

 In the end it's simple as adding a timeout the callback `onImageLoaded`... don't you think?

The only think that need to be 100% sure is that `onImageLoaded` will not fail at checking image is properly loaded.

Saying that because checking when the source of the image get switched from data-original to src, is not checking if the image is loaded.

**Comment by jimblue** [comment]
Lastly to prevent the problem of the first fetch you could create an empty `loaded` array that will be filled as `data-original` source are loaded:

```js
let loaded = []

if (loaded[imgIndex] === true) {
    onImageLoaded()
} else {
     fetchImage(imgSource).then(() => {
        onImageLoaded()
        loaded[imgIndex] = true
    })
}
```

This will help not impact performances.

**Comment by jimblue** [comment]
Hi @kingdido999!

Can you have your input on this please?

Tks
